spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -1419,6 +1477,57 @@ trait OptimisticTransactionImpl extends TransactionalWrite
     }
   }
 
+  /**
+   * This method registers the table with the CommitStore if the table is transitioning from
+   * file-system based table to managed-commit table.
+   * @param finalMetadata the effective [[Metadata]] of the table. Note that this refers to the
+   *                      new metadata if this commit is updating the table Metadata.
+   * @param finalProtocol the effective [[Protocol]] of the table. Note that this refers to the
+   *                      new protocol if this commit is updating the table Protocol.
+   * @return The new managed-commit table metadata if the table is transitioning from file-system
+   *         based table to managed-commit table. Otherwise, None.
+   *         This metadata should be added to the [[Metadata.configuration]] before doing the
+   *         commit.
+   */
+  protected def registerTableForManagedCommitsIfNeeded(
+      finalMetadata: Metadata,
+      finalProtocol: Protocol): Option[Map[String, String]] = {
+    val (oldOwnerName, oldOwnerConf) = CommitOwner.getManagedCommitConfs(snapshot.metadata)
+    var newManagedCommitTableConf: Option[Map[String, String]] = None
+    if (finalMetadata.configuration != snapshot.metadata.configuration || snapshot.version == -1L) {
+      val newCommitStoreOpt = CommitStoreProvider.getCommitStore(finalMetadata, finalProtocol)
+      (newCommitStoreOpt, readSnapshotTableCommitStoreOpt) match {
+        case (Some(newCommitStore), None) =>
+          // FS -> MC conversion
+          val (commitOwnerName, commitOwnerConf) = CommitOwner.getManagedCommitConfs(finalMetadata)
+          logInfo(s"Table ${deltaLog.logPath} transitioning from file-system based table to " +
+            s"managed-commit table: [commit owner: $commitOwnerName, conf: $commitOwnerConf]")
+          newManagedCommitTableConf = Some(
+            newCommitStore.registerTable(deltaLog.logPath, readVersion, finalMetadata, protocol))
+        case (None, Some(readCommitStore)) =>
+          // MC -> FS conversion
+          val (newOwnerName, newOwnerConf) = CommitOwner.getManagedCommitConfs(snapshot.metadata)
+          logInfo(s"Table ${deltaLog.logPath} transitioning from managed-commit table to " +
+            s"file-system table: [commit owner: $newOwnerName, conf: $newOwnerConf]")
+        case (Some(newCommitStore), Some(readCommitStore))
+            if !readCommitStore.semanticsEquals(newCommitStore) =>
+          // MC1 -> MC2 conversion is not allowed.
```

Comment:
Why are direct transitions between managed commit owners not allowed?

Reply:
The managed-commit spec forbids direct change of commit owners.